plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/alg/AlgorithmAssembler.java

```diff
@@ -0,0 +1,391 @@
+ * @see ILayoutProcessor
+ * @see ILayoutPhase
+ * @see LayoutProcessorConfiguration
+ */
+public final class AlgorithmAssembler<P extends Enum<P>, G>
+        extends AbstractRandomListAccessor<ILayoutPhaseFactory<P, G>> {
+
+    /** Whether layout processors and phases are cached once obtained from their factories. */
+    private boolean enableCaching = true;
+    /** Whether assembling the algorithm should fail if not all phases have implementations assigned. */
+    private boolean failOnMissingPhase = true;
+    /** The comparator used to sort processors in each processing slot. */
+    private Comparator<ILayoutProcessorFactory<G>> processorComparator = new EnumBasedFactoryComparator<>();
+
+    /** Class of the enumeration of all phases an algorithm is structured into. */
+    private final Class<P> phasesEnumClass;
+    /** The number of enumeration constants in the phase enumeration, saved for performance's sake. */
+    private final int numberOfPhases;
+    /** Set of phases whose implementations already have implementations assigned. */
+    private final Set<P> configuredPhases;
+    /** A configuration with additional processors which the phases don't necessarily depend upon themselves. */
+    private final LayoutProcessorConfiguration<P, G> additionalProcessors;
+    /** The cache for phase and processor instances, if caching is enabled. */
+    private final Map<ILayoutProcessorFactory<G>, ILayoutProcessor<G>> cache;
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // Creation
+
+    /**
+     * Creates a new algorithm assembler based on the given enumeration of phases the algorithm should be structured
+     * into.
+     * 
+     * @param phaseEnum
+     *            enumeration class.
+     * @return an algorithm assembler that can build algorithms structured into the given phases.
+     */
+    public static <P extends Enum<P>, G> AlgorithmAssembler<P, G> create(final Class<P> phaseEnum) {
+        return new AlgorithmAssembler<>(phaseEnum);
+    }
+
+    private AlgorithmAssembler(final Class<P> phaseEnum) {
+        phasesEnumClass = phaseEnum;
+        numberOfPhases = phaseEnum.getEnumConstants().length;
+
+        if (numberOfPhases == 0) {
+            throw new IllegalArgumentException("There must be at least one phase in the phase enumeration.");
+        }
+
+        configuredPhases = EnumSet.noneOf(phasesEnumClass);
+        additionalProcessors = LayoutProcessorConfiguration.create();
+        cache = Maps.newHashMap();
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // Configuration
+
+    /**
+     * Enables caching of layout phases and processors obtained from the factories. Caching is enabled by default.
+     * 
+     * @return this assembler to enable method chaining.
+     */
+    public AlgorithmAssembler<P, G> enableCaching() {
```

Comment:
I prefer a single method for this kind of configuration: `withCaching(bool)` or `caching(bool)`.